source/extensions/filters/http/jwt_authn/jwks_cache.cc

```diff
@@ -55,15 +61,36 @@ class JwksDataImpl : public JwksCache::JwksData, public Logger::Loggable<Logger:
     return audiences_->areAudiencesAllowed(jwt_audiences);
   }
 
-  const Jwks* getJwksObj() const override { return jwks_obj_.get(); }
+  const Jwks* getJwksObj() const override { return tls_->jwks_.get(); }
 
-  bool isExpired() const override { return time_source_.monotonicTime() >= expiration_time_; }
+  bool isExpired() const override { return time_source_.monotonicTime() >= tls_->expire_; }
 
   const ::google::jwt_verify::Jwks* setRemoteJwks(::google::jwt_verify::JwksPtr&& jwks) override {
-    return setKey(std::move(jwks), getRemoteJwksExpirationTime());
+    // convert unique_ptr to shared_ptr
+    JwksSharedPtr shared_jwks(jwks.release());
+    tls_->jwks_ = shared_jwks;
+    tls_->expire_ = getRemoteJwksExpirationTime();
+    return shared_jwks.get();
   }
 
 private:
+  struct ThreadLocalCache : public ThreadLocal::ThreadLocalObject {
+    // The jwks object.
+    JwksSharedPtr jwks_;
+    // The pubkey expiration time.
+    MonotonicTime expire_;
+  };
+
+  // Set jwks shared_ptr to all threads.
+  void setJwksToAllThreads(::google::jwt_verify::JwksPtr&& jwks,
+                           std::chrono::steady_clock::time_point expire) {
+    JwksSharedPtr shared_jwks(jwks.release());
```

Comment:
@qiwzhang how come this isn't a const shared pointer?

Reply:
It should be.  I will update it in the next pr. Thanks.